### PR TITLE
allow no-voters to withdraw

### PR DESCRIPTION
**Pull request by CJentzsch**
This is an alternative to https://github.com/slockit/DAO/pull/248 (only one should be merged)

### CODE REVIEW

File: DAO.sol [hunk 1]
```diff
@@ -615,6 +626,47 @@ contract DAO is DAOInterface, Token, TokenCreation {
...
+        // unVoteAll() would be an alternative
+        if (isSplitBlocked(msg.sender))
+            throw;
+
+        // Move ether
+        uint fundsToBeMoved =
+            (balances[msg.sender] * actualBalance()) /
+            totalSupply;
+
+        msg.sender.call.value(fundsToBeMoved);
```
Comment (LefterisJP) [inline DAO.sol]: **DANGER:** Here we are gonna become vulnerable to the recursive call exploit again! 

1. Change it to `send()`
2. Make all changes to the "state of the contract" before the actual send/call to protect from any recursive calls.

File: DAO.sol [hunk 2]
```diff
@@ -902,16 +954,28 @@ contract DAO is DAOInterface, Token, TokenCreation {
...
-        if (now > p.votingDeadline) {
+        if (p.open) {
             blocked[_account] = 0;
             return false;
         } else {
             return true;
         }
     }
 
+    function isSplitBlocked(address _account) internal returns (bool) {
```
Comment (LefterisJP) [inline DAO.sol]: Please change the name of this function here and wherever it is used. The function may also change the `splitBlocked[]` array. So a better name would be: `getOrModifySplitBlocked()`

File: DAO.sol [hunk 3]
```diff
@@ -902,16 +954,28 @@ contract DAO is DAOInterface, Token, TokenCreation {
...
+            splitBlocked[_account] = 0;
+            return false;
+        } else {
+            return true;
+        }
+    }
+
     function unblockMe() returns (bool) {
-        return isBlocked(msg.sender);
+        return isBlocked(msg.sender) && isSplitBlocked(msg.sender);
```
Comment (LefterisJP) [inline DAO.sol]: This is something was really bothering me with the v1.0 DAO. Please as part of this PR rename `isBlocked` to `getOrModifyBlocked()`

File: DAO.sol [hunk 4]
```diff
@@ -615,6 +627,45 @@ contract DAO is DAOInterface, Token, TokenCreation {
...
+        // don't allow withdrawing when token holder is blocked due to a vote
+        // unVoteAll() would be an alternative
+        if (getOrModifySplitBlocked(msg.sender))
+            throw;
+
+        // Move ether
+        uint senderBalance = balances[msg.sender];
+        uint fundsToBeMoved = (senderBalance * actualBalance()) / totalSupply;
+        balances[msg.sender] = 0;
+        msg.sender.send(fundsToBeMoved);
```
Comment (colm) [inline DAO.sol]: possible stack depth attack here. You need to wrap in if loop. I have fix but cannot push it to thedao repo.

File: DAO.sol [hunk 5]
```diff
@@ -890,28 +941,40 @@ contract DAO is DAOInterface, Token, TokenCreation {
...
         } else {
             return true;
         }
     }
 
+    function getOrModifySplitBlocked(address _account) internal returns (bool) {
+        if (splitBlocked[_account] == 0)
+            return false;
+        Proposal p = proposals[splitBlocked[_account]];
+        if (p.open) {
```
Comment (colm) [inline DAO.sol]: the token should stay blocked until the `splitExecutionPeriod` has passed.

Comment (LefterisJP) [inline DAO.sol]: `splitExecutionPeriod` is 27 days as can be seen [here](https://github.com/slockit/DAO/blob/master/DAO.sol#L37). Are you sure you mean that? Perhaps you mean the `splitGracePeriod` from [this](https://github.com/slockit/DAO/pull/246/files#diff-b30f9d3be2a832949bdd72e23956c130R44) PR.

If it's the latter `p.open` stays true until `executeProposal()` is succesfull, right? It's kind of the same check, only relying on someone having succesfully called `executeProposal()`. In the case of a faulty proposal which always throws, someone can always call `executeProposal()` to have it closed after the `waitingPeriod()` is over.

**Pull request merged**